Sources/NIOTransportServices/NIOTSConnectionChannel.swift

```diff
@@ -240,177 +251,32 @@ internal final class NIOTSConnectionChannel {
+        fatalError("option \(type(of: option)).\(option) not supported")
     }
 
-    public func setOption<Option: ChannelOption>(_ option: Option, value: Option.Value) -> EventLoopFuture<Void> {
-        if self.eventLoop.inEventLoop {
-            return self.eventLoop.makeCompletedFuture(Result { try setOption0(option: option, value: value) })
-        } else {
-            return self.eventLoop.submit { try self.setOption0(option: option, value: value) }
-        }
-    }
-
-    private func setOption0<Option: ChannelOption>(option: Option, value: Option.Value) throws {
+    func setChannelSpecificOption0<Option: ChannelOption>(option: Option, value: Option.Value) throws {
         self.eventLoop.preconditionInEventLoop()
 
         guard !self.closed else {
             throw ChannelError.ioOnClosedChannel
         }
 
         switch option {
-        case _ as ChannelOptions.Types.AutoReadOption:
-            self.options.autoRead = value as! Bool
-            self.readIfNeeded0()
-        case _ as ChannelOptions.Types.AllowRemoteHalfClosureOption:
-            self.options.supportRemoteHalfClosure = value as! Bool
-        case _ as ChannelOptions.Types.SocketOption:
-            let optionValue = option as! ChannelOptions.Types.SocketOption
-
-            // SO_REUSEADDR and SO_REUSEPORT are handled here.
-            switch (optionValue.level, optionValue.name) {
-            case (SOL_SOCKET, SO_REUSEADDR):
-                self.reuseAddress = (value as! SocketOptionValue) != Int32(0)
-            case (SOL_SOCKET, SO_REUSEPORT):
-                self.reusePort = (value as! SocketOptionValue) != Int32(0)
-            default:
-                try self.tcpOptions.applyChannelOption(option: optionValue, value: value as! SocketOptionValue)
-            }
-        case _ as ChannelOptions.Types.WriteBufferWaterMarkOption:
-            if self.backpressureManager.writabilityChanges(whenUpdatingWaterMarks: value as! ChannelOptions.Types.WriteBufferWaterMark) {
-                self.pipeline.fireChannelWritabilityChanged()
-            }
-        case _ as NIOTSChannelOptions.Types.NIOTSWaitForActivityOption:
-            let newValue = value as! Bool
-            self.options.waitForActivity = newValue
-
-            if let state = self.nwConnection?.state, case .waiting(let err) = state, !newValue {
-                // We're in waiting now, so we should drop the connection.
-                self.close0(error: err, mode: .all, promise: nil)
-            }
-        case is NIOTSChannelOptions.Types.NIOTSEnablePeerToPeerOption:
-            self.enablePeerToPeer = value as! NIOTSChannelOptions.Types.NIOTSEnablePeerToPeerOption.Value
-        case is NIOTSChannelOptions.Types.NIOTSAllowLocalEndpointReuse:
-            self.allowLocalEndpointReuse = value as! NIOTSChannelOptions.Types.NIOTSEnablePeerToPeerOption.Value
         case is NIOTSChannelOptions.Types.NIOTSMultipathOption:
```

Comment:
The `get` for this is in the `StateManagedNWConnectionChannel`, is there any reason the `set` shouldn't be?

Reply:
Ah, that's a mistake. The Multipart I believe should only be addd to TCP

Reply:
That was my intention anyhow